### PR TITLE
Adding functions to input custom image

### PR DESCRIPTION
**Pull request by tianhaoxie**
Hi, I combined part of the PTI code into the visualizer so that custom image can be uploaded directly. It add some code and makes some change to the viz/renderer. Maybe after testing, it can be added as a branch.

### CODE REVIEW

**Comment by radames** [comment]
@tianhaoxie very cool! how do we change the base model? is the default the horse one?

**Comment by tianhaoxie** [comment]
> @tianhaoxie very cool! how do we change the base model? is the default the horse one?



> @tianhaoxie very cool! how do we change the base model? is the default the horse one?

It follows the original way of DragGAN, default is lion and you can change it by changing checkpoints.

**Comment by radames** [comment]
sorry, I just notice that, testing! great it uses the selected checkpoint 👏 very cool!

**Comment by NandhaKishorM** [comment]
The branch is way behind the original, you have to update requirements stylegan_human directory

**Comment by NandhaKishorM** [comment]
It is not working with stylegan_human_v2_1024, tried  resize = (1024, 512) on L69 in inversion.py and also at L227     def train(self,img,w_plus=False):
        resize = (1024,512)
        if torch.is_tensor(img) == False:
            transform = transforms.Compose(
                [
                    transforms.Resize(resize, ),
                
                    transforms.ToTensor(),
                    transforms.Normalize([0.5, 0.5, 0.5], [0.5, 0.5, 0.5]),
                ]
            )
....

**Comment by tianhaoxie** [comment]
> It is not working with stylegan_human_v2_1024, tried resize = (1024, 512) on L69 in inversion.py and also at L227 def train(self,img,w_plus=False): resize = (1024,512) if torch.is_tensor(img) == False: transform = transforms.Compose( [ transforms.Resize(resize, ),
> 
> ```
>                 transforms.ToTensor(),
>                 transforms.Normalize([0.5, 0.5, 0.5], [0.5, 0.5, 0.5]),
>             ]
>         )
> ```
> 
> .... It will run, but wont make better results,

It only supports the pre-trained models that you can download in the script. Since the GAN inversion is based on PTI, I'm not sure if it supports stylegan_human.

**Comment by NandhaKishorM** [comment]
PTI do support in stylegan_human. https://github.com/stylegan-human/StyleGAN-Human

**Comment by radames** [comment]
hi @tianhaoxie , I also noticed that you might need to update the generator on the current model following this
https://github.com/danielroich/PTI/issues/26 wdyt?

```python
  tmp['G_ema'] = old_G.eval().requires_grad_(False).cpu()# copy.deepcopy(new_G).eval().requires_grad_(False).cpu()
  tmp['G'] = new_G.eval().requires_grad_(False).cpu() # copy.deepcopy(new_G).eval().requires_grad_(False).cpu()
  tmp['D'] = old_D
  tmp['training_set_kwargs'] = None
  tmp['augment_pipe'] = None
```

**Comment by 631068264** [comment]
I follow  the PTI and meet this https://github.com/danielroich/PTI/issues/50 , I hope this pr can help me.

**Comment by zhaoxiong123** [comment]
Hi, i got a picture when i run pti, but the picture is blurry。

please help me, thanks
@tianhaoxie

File: gan_inv/inversion.py [hunk 1]
```diff
@@ -0,0 +1,343 @@
...
+        optimizer = torch.optim.Adam(self.g_ema.parameters(), lr=self.pti_lr)
+        print('start PTI')
+        pbar = tqdm(range(self.max_pti_step))
+        for i in pbar:
+            t = i / self.max_pti_step
+            lr = get_lr(t, self.pti_lr)
+            optimizer.param_groups[0]["lr"] = lr
+
+            generated_image = self.g_ema.synthesis(ws,noise_mode='const')
+            loss = self.cacl_loss(self.percept,generated_image,real_img)
```
[the remark this answers is not in the document]
Comment (tianhaoxie) [reply gan_inv/inversion.py:266]: yeah, stylegan2 human is different from others, which has 1024by512 resolution, whereas typically the H and W are same.

File: gan_inv/__pycache__/PTI.cpython-39.pyc
[the remark this answers is not in the document]
Comment (tianhaoxie) [reply gan_inv/__pycache__/PTI.cpython-39.pyc:1]: Thanks for your suggestion!

**Comment by tianhaoxie** [comment]
> I follow the PTI and meet this [danielroich/PTI#50](https://github.com/danielroich/PTI/issues/50) , I hope this pr can help me.

It's because the image resolution is depended on the stylegan checkpoint you selected. If the image you input is not exactly the resolution of the checkpoint, you need to crop and resize.

**Comment by tianhaoxie** [comment]
> Hi, i got a picture when i run pti, but the picture is blurry。  please help me, thanks @tianhaoxie

It is normal for inversion to get results blurred and distorted.

**Comment by marii89** [comment]
is there a possibility that in the future we can edit any image we add?

**Comment by MittalNeha** [comment]
@tianhaoxie I am trying to run the gradio visualizer from your github but it is not working! Can you possibly lay down the steps you followed to generate the gan_inv checkpoints and what all are the limitations for running this demo? 
When I load the custom image, the gradio UI is stuck somewhere.

**Comment by tianhaoxie** [comment]
> @tianhaoxie I am trying to run the gradio visualizer from your github but it is not working! Can you possibly lay down the steps you followed to generate the gan_inv checkpoints and what all are the limitations for running this demo? When I load the custom image, the gradio UI is stuck somewhere.

Except for the stylegan human, all other checkpoints should work. Btw, you should upload your image by clicking 'inverse custom image', rather than drop it in the image box.

**Comment by marii89** [comment]
> > @tianhaoxie I am trying to run the gradio visualizer from your github but it is not working! Can you possibly lay down the steps you followed to generate the gan_inv checkpoints and what all are the limitations for running this demo? When I load the custom image, the gradio UI is stuck somewhere.
> 
> Except for the stylegan human, all other checkpoints should work. Btw, you should upload your image by clicking 'inverse custom image', rather than drop it in the image box.

we have feature in draggan for edit any uploaded image?

**Comment by tianhaoxie** [comment]
> > > @tianhaoxie I am trying to run the gradio visualizer from your github but it is not working! Can you possibly lay down the steps you followed to generate the gan_inv checkpoints and what all are the limitations for running this demo? When I load the custom image, the gradio UI is stuck somewhere.
> > 
> > 
> > Except for the stylegan human, all other checkpoints should work. Btw, you should upload your image by clicking 'inverse custom image', rather than drop it in the image box.
> 
> we have feature in draggan for edit any uploaded image?

no, only the categories that the pre-trained checkpoints we have(i.e. horse, face).

**Comment by marii89** [comment]
> > > > @tianhaoxie I am trying to run the gradio visualizer from your github but it is not working! Can you possibly lay down the steps you followed to generate the gan_inv checkpoints and what all are the limitations for running this demo? When I load the custom image, the gradio UI is stuck somewhere.
> > > 
> > > 
> > > Except for the stylegan human, all other checkpoints should work. Btw, you should upload your image by clicking 'inverse custom image', rather than drop it in the image box.
> > 
> > 
> > we have feature in draggan for edit any uploaded image?
> 
> no, only the categories that the pre-trained checkpoints we have(i.e. horse, face).

if i good understand you have pre-trained categories e.g. face, horse.. and i can upload any photo where is face e.g. face of my sister and edit with draggan but i can't upload some category which is not trained e.g. cat?

File: gan_inv/inversion.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (sukablyadskaya) [reply gan_inv/inversion.py:266]: hello can i train images on ffhq 512 not 1024 because it's too slow in Draggan

**Comment by JanineCHEN** [comment]
> > > > > @tianhaoxie I am trying to run the gradio visualizer from your github but it is not working! Can you possibly lay down the steps you followed to generate the gan_inv checkpoints and what all are the limitations for running this demo? When I load the custom image, the gradio UI is stuck somewhere.
> > > > 
> > > > 
> > > > Except for the stylegan human, all other checkpoints should work. Btw, you should upload your image by clicking 'inverse custom image', rather than drop it in the image box.
> > > 
> > > 
> > > we have feature in draggan for edit any uploaded image?
> > 
> > 
> > no, only the categories that the pre-trained checkpoints we have(i.e. horse, face).
> 
> if i good understand you have pre-trained categories e.g. face, horse.. and i can upload any photo where is face e.g. face of my sister and edit with draggan but i can't upload some category which is not trained e.g. cat?

I believe so, the provided image should roughly belong to a data distribution approximated by any of the provided trained stylegan models.

**Comment by timchenxiaoyu** [comment]
looking forward
